### PR TITLE
Add exhaustive test for group functions on a low-order subgroup

### PR DESCRIPTION
**Pull request by apoelstra**
We observe that when changing the b-value in the elliptic curve formula
`y^2 = x^3 + ax + b`, the group law is unchanged. Therefore our functions
for secp256k1 will be correct if and only if they are correct when applied
to the curve defined by `y^2 = x^3 + 4` defined over the same field. This
curve has a point P of order 199.

This commit adds a test which computes the subgroup generated by P and
exhaustively checks that addition of every pair of points gives the correct
result.

Unfortunately we cannot test scalar multiplication, const-time or otherwise,
by the same mechanism. The reason is that our ecmult functions both compute
a wNAF representation of the scalar, and this representation is tied to the
order of the group.

Testing with the incomplete version of gej_add_ge (found in 5de4c5dff^)
shows that this detects the incompleteness when adding P - 106P, which
is exactly what we expected since 106 is a cube root of 1 mod 199.

Exhaustive tests added for the following internal functions:
- `secp256k1_ge_is_infinity`, `secp256k1_gej_is_infinity`
- `secp256k1_gej_add_var`, `secp256k1_gej_add_ge`, `secp256k1_gej_add_ge_var`, `secp256k1_gej_add_zinv_var`
- `secp256k1_gej_double_nonzero`, `secp256k1_gej_double_var`
- `secp256k1_ge_neg`, `secp256k1_gej_neg`
- `secp256k1_ecmult`, `secp256k1_ecmult_const`

And also (unfortunately with a bit of surgery; the original functions don't work with small group orders)
- `secp256k1_ecdsa_sign`
- `secp256k1_ecdsa_verify`

### CODE REVIEW

**Comment by apoelstra** [comment]
Fixes #308.

**Comment by apoelstra** [comment]
Pushed a bounds-check fix that @gmaxwell pointed out (so we no longer test negation of infinity). Also added a test for negating infinity elsewhere in the code so that this edge case is tested somewhere.

The problem with testing -infinity along with the other loworder stuff was that setting an unitialized point to infinity leaves its y-coordinate uninitialized, and when calling fe_negate on this, FE_VERIFY branches on it. The result is a valgrind error, and the result of some discussion on IRC was that it'd be best to explicitly test negating fully-initialized infinities elsewhere in tests.c.

File: src/tests.c [hunk 1]
```diff
@@ -1271,6 +1283,95 @@ void run_ge(void) {
...
+    /* Generate the entire group */
+    secp256k1_ge_set_infinity(&group[0]);
+    secp256k1_gej_set_infinity(&groupj[0]);
+    for (i = 1; i < order; i++) {
+        secp256k1_fe_t z;
+        random_fe(&z);
+
+        secp256k1_gej_add_ge(&groupj[i], &groupj[i - 1], &gen);
+        secp256k1_ge_set_gej(&group[i], &groupj[i]);
+        secp256k1_fe_mul(&groupj[i].z, &groupj[i].z, &z);
```
Comment (sipa) [inline src/tests.c]: Use secp256k1_gej_rescale ?

**Comment by apoelstra** [comment]
Push to move exhaustive tests to their own binary; add testing for non-constant ecmult code.

**Comment by sipa** [comment]
Needs rebase (even if github doesn't realize that).

**Comment by apoelstra** [comment]
Rebased and cleaned up. I think this is ready for review.

There aren't any exhaustive tests for modules yet.

**Comment by sipa** [comment]
Travis's on fire, yo.

**Comment by apoelstra** [comment]
Rebased and completed verification

**Comment by apoelstra** [comment]
Updated to just disable static precomputation for `tests_exhaustive.c` instead of whining to the user about her ./configure options.

File: src/ecdsa_impl.h [hunk 2]
```diff
@@ -203,7 +203,9 @@ static int secp256k1_ecdsa_sig_serialize(unsigned char *sig, size_t *size, const
 static int secp256k1_ecdsa_sig_verify(const secp256k1_ecmult_context *ctx, const secp256k1_scalar *sigr, const secp256k1_scalar *sigs, const secp256k1_ge *pubkey, const secp256k1_scalar *message) {
     unsigned char c[32];
     secp256k1_scalar sn, u1, u2;
+#if !defined(EXHAUSTIVE_TEST_ORDER)
```
Comment (sipa) [inline src/ecdsa_impl.h]: ifndef ?

File: src/scalar_low.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/**********************************************************************
+ * Copyright (c) 2014 Pieter Wuille                                   *
```
Comment (sipa) [inline src/scalar_low.h]: I can't remember writing this code!

File: src/scalar_low.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+#ifndef _SECP256K1_SCALAR_REPR_
+#define _SECP256K1_SCALAR_REPR_
+
+#include <stdint.h>
+
+/** A scalar modulo the group order of the secp256k1 curve. */
+typedef uint32_t secp256k1_scalar;
+
+#define SECP256K1_SCALAR_CONST(d7, d6, d5, d4, d3, d2, d1, d0) ((uint32_t)(d0))
```
Comment (sipa) [inline src/scalar_low.h]: Is this called anywhere? If it's needed, maybe it's better to have a SECP256K1_SCALAR_CONST_SMALL instead that just takes a single uint32_t, and in the non-low case call the normal one?

File: src/ecdsa_impl.h (hunk 2 above)
Comment (apoelstra) [inline src/ecdsa_impl.h]: We're inconsistent about this throughout the code. Unless you feel strongly about this I'm not going to change it .. we can do a cleanup pass at some point in the future.

Comment (sipa) [inline src/ecdsa_impl.h]: OKAY

File: src/ecmult_const_impl.h [hunk 5]
```diff
@@ -78,7 +78,7 @@ static int secp256k1_wnaf_const(int *wnaf, secp256k1_scalar s, int w) {
     /* Negative numbers will be negated to keep their bit representation below the maximum width */
     flip = secp256k1_scalar_is_high(&s);
     /* We add 1 to even numbers, 2 to odd ones, noting that negation flips parity */
-    bit = flip ^ (s.d[0] & 1);
+    bit = flip ^ !secp256k1_scalar_is_even(&s);
```
Comment (sipa) [inline src/ecmult_const_impl.h]: Good. Layer violation prevented.

File: src/scalar_low_impl.h [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+SECP256K1_INLINE static int secp256k1_scalar_is_even(const secp256k1_scalar *a) {
+    return !(*a & 1);
+}
+
+SECP256K1_INLINE static void secp256k1_scalar_clear(secp256k1_scalar *r) { *r = 0; }
+SECP256K1_INLINE static void secp256k1_scalar_set_int(secp256k1_scalar *r, unsigned int v) { *r = v; }
+
+SECP256K1_INLINE static unsigned int secp256k1_scalar_get_bits(const secp256k1_scalar *a, unsigned int offset, unsigned int count) {
+    if (offset < 32)
```
Comment (sipa) [inline src/scalar_low_impl.h]: Either the if/else is unnecessary, or the `(offset < 32) * ` is.

File: src/scalar_low_impl.h [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        return (offset < 32) * ((*a >> offset) & ((((uint32_t)1) << count) - 1));
+    else
+        return 0;
+}
+
+SECP256K1_INLINE static unsigned int secp256k1_scalar_get_bits_var(const secp256k1_scalar *a, unsigned int offset, unsigned int count) {
+    if (offset < 32)
+        return (offset < 32) * ((*a >> offset) & ((((uint32_t)1) << count) - 1));
```
Comment (sipa) [inline src/scalar_low_impl.h]: Same here.

**Comment by apoelstra** [comment]
Updated to force-disable the endomorphism optimization for the exhaustive tests, since the `scalar_low` impl it uses does not support it.

**Comment by apoelstra** [comment]
Updated again to put endomorphism support back and just implement `SECP256K1_SCALAR_CONST`, which was the point of incompatibility. Turns out the Schnorr module also needs this macro.

**Comment by sipa** [comment]
Hmm, how can the endomorphism work without changing the lambda constant? (a cubic root of 1 mod the curve order)

**Comment by apoelstra** [comment]
@sipa it observably does .. I have not checked into why but my belief is that it always splits `n` into `n + 0*lambda` for such small `n`.

File: src/scalar_low_impl.h [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+SECP256K1_INLINE static void secp256k1_scalar_set_int(secp256k1_scalar *r, unsigned int v) { *r = v; }
+
+SECP256K1_INLINE static unsigned int secp256k1_scalar_get_bits(const secp256k1_scalar *a, unsigned int offset, unsigned int count) {
+    if (offset < 32)
+        return ((*a >> offset) & ((((uint32_t)1) << count) - 1));
+    else
+        return 0;
+}
+
+SECP256K1_INLINE static unsigned int secp256k1_scalar_get_bits_var(const secp256k1_scalar *a, unsigned int offset, unsigned int count) {
```
Comment (sipa) [inline src/scalar_low_impl.h]: Does this need a separate implementation from secp256k1_scalar_get_bits? It could just call the other function, or even be #defined to be identical.

File: src/scalar_impl.h [hunk 9]
```diff
@@ -259,6 +281,18 @@ static void secp256k1_scalar_inverse_var(secp256k1_scalar *r, const secp256k1_sc
...
+#if defined(EXHAUSTIVE_TEST_ORDER)
+/**
+ * Find k1 and k2 given k, such that k1 + k2 * lambda == k mod n; unlike in the
+ * full case we don't bother making k1 and k2 be small, we just want them to be
+ * nontrivial to get full test coverage for the exhaustive tests. We therefore
+ * (arbitrarily) set k2 = k + 5 and k1 = k - k2 * lambda.
+ */
+static void secp256k1_scalar_split_lambda(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *a) {
+    *r2 = (*a + 5) % EXHAUSTIVE_TEST_ORDER;
+    *r1 = (10000 * EXHAUSTIVE_TEST_ORDER + *a - *r2 * EXHAUSTIVE_TEST_LAMBDA) % EXHAUSTIVE_TEST_ORDER;
```
Comment (sipa) [inline src/scalar_impl.h]: I think this can be written as `r1 = (*a + (EXHAUSTIVE_TEST_ORDER - *r2) * EXHAUSTIVE_TEST_LAMBDA) & EXHAUSTIVE_TEST_ORDER;`, avoiding the weird constant.

File: src/scalar_low_impl.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    if (offset < 32)
+        return ((*a >> offset) & ((((uint32_t)1) << count) - 1));
+    else
+        return 0;
+}
+
+SECP256K1_INLINE static int secp256k1_scalar_check_overflow(const secp256k1_scalar *a) { return *a > EXHAUSTIVE_TEST_ORDER; }
```
Comment (sipa) [inline src/scalar_low_impl.h]: `>=` ?

File: src/scalar_low_impl.h [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    bin[28] = *a >> 24; bin[29] = *a >> 16; bin[30] = *a >> 8; bin[31] = *a;
+}
+
+SECP256K1_INLINE static int secp256k1_scalar_is_zero(const secp256k1_scalar *a) {
+    return *a == 0;
+}
+
+static void secp256k1_scalar_negate(secp256k1_scalar *r, const secp256k1_scalar *a) {
+    *r = EXHAUSTIVE_TEST_ORDER - *a;
+    *r %= EXHAUSTIVE_TEST_ORDER;
```
Comment (sipa) [inline src/scalar_low_impl.h]: Faster: `if (*a) { *r = EXHAUSTIVE_TEST_ORDER - *a; } else { *r = 0; }`.

File: src/tests_exhaustive.c [hunk 12]
```diff
@@ -120,26 +155,90 @@ void test_exhaustive_addition(const secp256k1_ge *group, const secp256k1_gej *gr
...
+    secp256k1_fe x;
+    unsigned char x_bin[32];
+    k %= EXHAUSTIVE_TEST_ORDER;
+    x = group[k].x;
+    secp256k1_fe_normalize(&x);
+    secp256k1_fe_get_b32(x_bin, &x);
+    secp256k1_scalar_set_b32(r, x_bin, NULL);
+}
+
+/* hee hee hee */
```
Comment (sipa) [inline src/tests_exhaustive.c]: LOL

File: src/tests_exhaustive.c [hunk 13]
```diff
@@ -120,26 +155,90 @@ void test_exhaustive_addition(const secp256k1_ge *group, const secp256k1_gej *gr
...
+        secp256k1_scalar check_x;
+        r_from_k(&check_x, group, i);
+        if (*x_coord == check_x) {
+           return i;
+        }
+    }
+    return -1;
+}
+
+void test_exhaustive_sign(const secp256k1_context *ctx, const secp256k1_ge *group, int order) {
```
Comment (sipa) [inline src/tests_exhaustive.c]: Is there a use in passing order as a parameter? Almost all of this code already depends on knowing it is equal to EXHAUSTIVE_TEST_ORDER anyway.

File: src/scalar_impl.h [hunk 14]
```diff
@@ -259,6 +281,18 @@ static void secp256k1_scalar_inverse_var(secp256k1_scalar *r, const secp256k1_sc
...
+#if defined(EXHAUSTIVE_TEST_ORDER)
+/**
+ * Find k1 and k2 given k, such that k1 + k2 * lambda == k mod n; unlike in the
+ * full case we don't bother making k1 and k2 be small, we just want them to be
+ * nontrivial to get full test coverage for the exhaustive tests. We therefore
+ * (arbitrarily) set k2 = k + 5 and k1 = k - k2 * lambda.
+ */
+static void secp256k1_scalar_split_lambda(secp256k1_scalar *r1, secp256k1_scalar *r2, const secp256k1_scalar *a) {
+    *r2 = (*a + 5) % EXHAUSTIVE_TEST_ORDER;
+    *r1 = (*a + (EXHAUSTIVE_TEST_ORDER - *r2) * EXHAUSTIVE_TEST_LAMBDA) % EXHAUSTIVE_TEST_ORDER;;;;
```
Comment (sipa) [inline src/scalar_impl.h]: ;;;;?

File: src/scalar_low_impl.h (hunk 8 above)
Comment (sipa) [inline src/scalar_low_impl.h]: Fixed.
